### ISSUE TITLE
Use of maintenance_off / maintenance_on files

### ISSUE THREAD

**Issue by jpmjpmjpm** [jcalazan/ansible-django-stack#153]
I'm having hard time to understand the way `maintenance_xxx.html` files are used.

On the one hand `maintenance_off.html` is created in [roles/web/tasks/setup_virtualenv.yml](https://github.com/jcalazan/ansible-django-stack/blob/137358658e6314597cf1772ca11703b8f791fcd0/roles/web/tasks/setup_virtualenv.yml), but never referenced nor deleted.

On the other hand, the only reference to a maintenance file is in [roles/nginx/templates/youtubeadl.j2](https://github.com/jcalazan/ansible-django-stack/blob/206232ee417865706042eecf70ed8f3477dc4a68/roles/nginx/templates/youtubeadl.j2), the nginx template. But it is for `maintenance_on.html` file which is never created.

**In short I don't understand how the maintenance operation is working.** Even more, is this operation really performed and working? I have doubts as I was able to access my site while the ansible playbook was running.

Thanks for shading light on the topic!

**Comment by DavidCain** [comment]
To my understanding, `maintenance_on.html` is intended to be created *manually* if you ever need to do anything that would require downtime. When the file exists, NGINX will not serve any traffic to your Django application, and will instead serve just that static HTML. When you remove the file, NGINX will direct traffic back to your Django application.

The system has actually existed this way (unmodified) since the [first non-README commit](https://github.com/jcalazan/ansible-django-stack/commit/d5ba25f6046bd9e5f41686e9282a98491eac253b)!

One shouldn't generally have to turn maintenance mode on when running this playbook - operations such as changing dependencies, updating code, running migrations, and more can all be executed safely while the old gunicorn workers continue to serve traffic (and restarting gunicorn will use the new virtual environment & code when ready). Some database migrations might break without newer application code, and that's a real use case to turn maintenance mode on.

-----

(For what it's worth, I actually tried to make use of `maintenance_on.html` a couple times, but it never actually redirected traffic properly). I generally do any breaking maintenance in the wee hours of the morning, when I'm positive that no users will be hitting the system (& I just settle for ugly 500 errors if they do).

**Comment by YPCrumble** [comment]
@jpmjpmjpm @DavidCain I believe the `off` portion of `maintenance_off.html` is actually supposed to be the "flag" but it's a little confusing for sure. 

To put my sites in maintenance mode I just run something like `mv maintenance_off.html maintenance_on.html` and the maintenance page shows up. I just tested on one of my servers and it worked as expected - no need to restart NGINX or Gunicorn or anything. 

The output looks like the image below - remember to change the "contact us" email in the template:

![image](https://user-images.githubusercontent.com/12019216/85322113-c5cd9c00-b493-11ea-8058-55ed9b4b90f1.png)

@jpmjpmjpm if you wanted to add PR to document this better in the README it would be very much appreciated!

**Comment by DavidCain** [comment]
Yup, @YPCrumble - I definitely get that the `off` portion is meant to be the flag - the file does nothing when named `off`, but is meant to trigger maintenance mode when it's renamed to be `on`.

I haven't tried using it for a couple years now, so glad to see that it can work! I honestly never spent much time trying to make it work apart from that `mv` command & restarting NGINX for good measure.

It's definitely not well-documented now, so I'd love to see a quick README note about how to use it, or even an optional Ansible task to turn on maintenance mode.

**Comment by jpmjpmjpm** [comment]
@DavidCain @YPCrumble Thank you for your precious comments.

@DavidCain I'm probably missing something David! When you say *and more can all be executed safely while the old gunicorn workers continue to serve traffic (and restarting gunicorn will use the new virtual environment & code when ready).*

Should I understand that when using the playbook to update a server, **I should use a virtualenv with a different name than the previous one?** In order as you said to have the old unicorn worker continue to serve the traffic?

And I was indeed having in mind to update the playbook in order to have the switch activated automatically. In particular as I was fearing that during models migration, the code and the DB would be out of sync.

*As a totally side topic... I'm having hard time to figure out how to ssh the `application name` user in the vagrant box. How are you doing that? I'm trying to use [this post](https://stackoverflow.com/questions/22643177/ssh-onto-vagrant-box-with-different-username) as a help... but with limited success so far* Sorry to waste some of your time of that system topic.

Thanks to you both.

**Comment by DavidCain** [comment]
Sorry, I probably added more complication there than necessary.

No, you don't need to anything special with virtual environments! All I mean to say is that it's safe to change your Python source code - it won't affect the running application. Similarly, you can add/change any dependencies directly within the virtual environment and it won't affect the running application. The gunicorn workers won't use the new code or new dependencies until you explicitly restart the application. The Ansible playbook is accordingly set up to do lots of work like running migrations or pulling new code, _then_ restarting the gunicorn workers.

The relevance of the above statements are to say that using "maintenance mode" isn't really required in most cases. You can simply run the playbook, let it automatically pull the latest code from your Git remote & update any dependencies, and then eventually have it restart your gunicorn workers. Your users won't see any interruption.

Only one user can SSH into the server. A plain `vagrant ssh` is all you need to SSH in, then a `sudo su <application_name>` will let you change to that user.

For security reasons, it's never a good idea to have your application's user able to SSH into the box! Consider it a feature that only one user may SSH in. =)

**Comment by jpmjpmjpm** [comment]
@DavidCain Thanks David for those precision's. I have to learn more on `gunicorn` as I wasn't aware of this 'code caching mechanism.
There is still one point that worries me: *db update during django migrate.* Shouldn't maintenance mode be raised during this phase? If not, can't you have user requests incoming on a partially updated db or on a fully updated db, but with code not updated yet?

@YPCrumble Yes, I'll be happy to update README about current maintenance mechanism. Just let me some days to make it happen!

**Comment by DavidCain** [comment]
Yes, without having "maintenance mode" on, you can absolutely have requests coming in while the database is being migrated. That's _okay_ - modern databases are designed so that this scenario is perfectly acceptable. Django sites everywhere (and, really, all database-backed web applications) regularly perform database migrations with zero interruption to provided services.

Databases like Postgres have transactional DDL - so if there's any issue with the migration, rolling it back is simple and your application will never notice any interruption.

It's a little out of scope for this repository, but the trick to maintaining full uptime (while database changes are in-progress, or already applied) is to make sure that your current application can work with the future database changes. If you make sure about that, then you can easily run migrations on the existing application, _then_  release the next version of your application.

For example, you could:
1. Modify your Django models to add a new column
2. Run the migration that adds that new column to your table (ideally with a default, or a nullable value)
3. Be content that your current application won't look at the column when using the ORM, but it can still read and write rows from the table.
4. Deploy a newer bit of source code that now makes use of the column.

If you want to do something larger than a migration, like:
- totally changing the way that an in-use table works (in a backwards-incompatible way)
- updating the database software (e.g. switching to a new major version of Postgres)

Then yes, you might want to use maintenance mode.
But the fact of the matter is that relational databases (in fact, most all databases) are designed to be modified while in-use!

**Comment by jpmjpmjpm** [comment]
@DavidCain Thanks David for those additional elements.

What you describe below

> It's a little out of scope for this repository, but the trick to maintaining full uptime (while database changes are in-progress, or already applied) is to make sure that your current application can work with the future database changes. If you make sure about that, then you can easily run migrations on the existing application, then release the next version of your application.

is close to what is called blue-green deployment if I understand well!

**Comment by DavidCain** [comment]
Somewhat related, sure!

For what it's worth, this repository doesn't support an out-of-the-box configuration for true blue-green (running two production environments). It's meant to be a simple starting point that you could use to expand upon as needed.
